cpp/include/rapidsmpf/buffer/pinned_memory_resource.hpp

```diff
@@ -0,0 +1,236 @@
+ * host memory pool. This internally uses
+ * `cuda::experimental::pinned_memory_resource`. See for more details:
+ * https://nvidia.github.io/cccl/cudax/api/classcuda_1_1experimental_1_1pinned__memory__resource.html
+ *
+ * This class provides an interface for allocating and deallocating pinned
+ * (page-locked) host memory asynchronously using CUDA streams.
+ */
+class PinnedMemoryResource {
+    friend class PinnedHostBuffer;
+
+  public:
+    /**
+     * @brief Constructs a new pinned memory resource.
+     *
+     * @param pool The pinned memory pool to use for allocations.
+     */
+    PinnedMemoryResource(PinnedMemoryPool& pool);
+
+    /**
+     * @brief Destroys the pinned memory resource.
+     *
+     * Note: This does not deallocate memory that was allocated through this resource.
+     * All allocated memory should be explicitly deallocated before destruction.
+     */
+    ~PinnedMemoryResource();
+
+    /**
+     * @brief Allocates pinned memory asynchronously.
+     *
+     * @param bytes The number of bytes to allocate.
+     * @param stream_ref The CUDA stream to use for the allocation operation.
+     * @return A pointer to the allocated memory, or nullptr if allocation failed.
+     */
+    void* allocate_async(size_t bytes, const cuda::stream_ref stream_ref);
+
+    /**
+     * @brief Deallocates pinned memory asynchronously.
+     *
+     * @param ptr A pointer to the memory to deallocate.
+     * @param stream_ref The CUDA stream to use for the deallocation operation.
+     */
+    void deallocate_async(void* ptr, const cuda::stream_ref stream_ref);
+
+  private:
+    struct PinnedMemoryResourceImpl;
+    std::unique_ptr<PinnedMemoryResourceImpl> impl_;
+};
+
+/**
+ * @brief A buffer that manages pinned host memory.
+ *
+ * This class provides a convenient interface for managing pinned host memory
+ * buffers. It automatically handles allocation and deallocation through a
+ * PinnedMemoryResource and supports both empty buffer creation and buffer
+ * creation with data copying.
+ */
+class PinnedHostBuffer {
+  public:
+    /**
+     * @brief Constructs an empty pinned host buffer.
+     *
+     * @param size The size of the buffer in bytes.
+     * @param stream The CUDA stream to use for memory operations.
+     * @param p_resource Pointer to the memory resource to use for allocation and
+     * deallocation.
+     */
+    PinnedHostBuffer(
+        size_t size, cuda::stream_ref stream, PinnedMemoryResource* p_resource
+    );
+
+    /**
+     * @brief Constructs a pinned host buffer and copies data into it.
+     *
+     * @param data Pointer to the source data to copy.
+     * @param size The size of the data to copy in bytes.
+     * @param stream The CUDA stream to use for memory operations.
+     * @param p_resource Pointer to the memory resource to use for allocation and
+     * deallocation.
+     */
+    PinnedHostBuffer(
+        void const* data,
+        size_t size,
+        cuda::stream_ref stream,
```

Comment:
Let's be consistent, either `rmm::cuda_stream_view` or `cuda::stream_ref` everywhere 
```suggestion
        rmm::cuda_stream_view stream,
```

Reply:
I'm more inclined to use `cuda::stream_ref ` because we are not using any rmm dependencies in this class. rmm stream view is implicitly convertible to cuda stream ref. So, it won't be a big issue.

Reply:
Is stream_ref implicitly convertible to cuda_stream_view? If not that is painful with the rest of rapidsmpf